### PR TITLE
feat: Add debug functions for transport (`onRequest`, `onResponse`)

### PR DESCRIPTION
**Pull request by insulineru**
---

### Focus of this PR:
Add `onRequest` and `onResponse` helpers for transport debug.

### Detailed summary:
- Added `onRequest` and `onResponse` helpers for transport debug in `playgrounds/browser/src/components/clients/HttpPublic.tsx`, `src/clients/transports/fallback.ts`, `src/clients/transports/http.ts`, `src/clients/transports/custom.ts`, `src/utils/buildRequest.ts`, `src/clients/transports/createTransport.test.ts`, `site/docs/clients/transports/custom.md`, `site/docs/clients/transports/http.md`, and `site/docs/clients/transports/fallback.md`.

> The following files were skipped due to too many changes: `site/docs/clients/transports/fallback.md`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/clients/transports/createTransport.ts [hunk 1]
```diff
@@ -13,6 +14,10 @@ export type TransportConfig<
   key: string
   /** The JSON-RPC request function that matches the EIP-1193 request spec. */
   request: TEIP1193RequestFn
+  /** On request */
+  onRequest?: (request: RpcRequest[]) => void
+  /** On response */
+  onResponse?: (response: any) => void
```
Comment (tmm) [inline src/clients/transports/createTransport.ts:20]:
```suggestion
  onResponse?: (response: unknown) => void
```

File: src/clients/transports/http.ts [hunk 2]
```diff
@@ -37,6 +37,10 @@ export type HttpTransportConfig = {
   retryDelay?: TransportConfig['retryDelay']
   /** The timeout (in ms) for the HTTP request. Default: 10_000 */
   timeout?: TransportConfig['timeout']
+  /** On request debug helper */
+  onRequest?: (request: RpcRequest[]) => void
+  /** On response */
+  onResponse?: (response: any) => void
```
Comment (tmm) [inline src/clients/transports/http.ts:43]:
```suggestion
  onResponse?: (response: unknown) => void
```

**Review by tmm — COMMENTED** [review]
Looking good. Type-level feedback.

File: src/clients/transports/createTransport.ts (hunk 1 above)
Comment (insulineru) [reply src/clients/transports/createTransport.ts]: Refactored here, thanks for the comment

File: src/clients/transports/http.ts (hunk 2 above)
Comment (insulineru) [reply src/clients/transports/http.ts]: I made a little mistake here, as I didn't reuse `TransportConfig` as I did in other transports. Corrected :)

**Comment by insulineru** [comment]
GM, guys. Could you please take a look at Pull Request in your spare time? @tmm @jxom

**Comment by insulineru** [comment]
GM, guys. Could you please take a look at Pull Request in your spare time? @tmm @jxom

**Review by tmm — COMMENTED** [review]
Do we want to expose a `onError` property as well?

File: site/docs/clients/transports/custom.md [hunk 3]
```diff
@@ -131,6 +131,32 @@ const transport = custom(window.ethereum, {
...
+
+```ts
+const transport = webSocket('wss://eth-mainnet.g.alchemy.com/v2/...', {
+  onRequest: (request) => console.log(request), // [!code focus]
+})
+```
+
+### onResponse (optional)
+
+- **Type:** `(response: unknown) => void`
```
Comment (jxom) [inline site/docs/clients/transports/custom.md:149]: Would be nice if `onResponse` followed the shape of Fallback Transport's `onResponse`: https://github.com/wagmi-dev/viem/blob/main/src/clients/transports/fallback.ts#L11-L28

File: site/docs/clients/transports/custom.md [hunk 4]
```diff
@@ -131,6 +131,32 @@ const transport = custom(window.ethereum, {
 })
 ```
 
+### onRequest (optional)
+
+- **Type:** `(request: RpcRequest[]) => void`
```
Comment (jxom) [inline site/docs/clients/transports/custom.md:136]: Isn't this meant to be `RpcRequest`?
```suggestion
- **Type:** `(request: RpcRequest) => void`
```

Comment (jxom) [reply site/docs/clients/transports/custom.md:136]: Could be nice to also include `transport` here too.

**Comment by jxom** [comment]
@insulineru – do you need any help with this?

**Comment by insulineru** [comment]
@jxom sorry, i skipped comments. I'll try to make changes in the next couple days, thanks for the review

File: site/docs/clients/transports/http.md [hunk 5]
```diff
@@ -202,3 +202,28 @@ const transport = http('https://eth-mainnet.g.alchemy.com/v2/...', {
...
 
+### onRequest (optional)
+
+- **Type:** `(request: RpcRequest[]) => void`
+- **Default:** `undefined`
+
+A callback that is called when a request is sent.
+
+```ts
+const transport = webSocket('wss://eth-mainnet.g.alchemy.com/v2/...', {
```
Comment (holic) [inline site/docs/clients/transports/http.md:213]: should the examples here use `http` not `webSocket`?

**Comment by holic** [comment]
Another approach that might be useful is something like

```
onRequest: (request: RpcRequest, response: Promise<unknown>) => void;
```

so users of this callback can tie together the request and its response without having to figure this out downstream using both `onRequest`/`onResponse` callbacks (sounds tricky!)

**Comment by tmm** [comment]
@holic This seems more intuitive to me since `onResponse` comes after `onRequest`.

```ts
onRequest(request: RpcRequest): void
onResponse(response: Promise<unknown>, request: RpcRequest): void
```

**Comment by holic** [comment]
> @holic This seems more intuitive to me since `onResponse` comes after `onRequest`.
> 
> ```ts
> onRequest(request: RpcRequest): void
> onResponse(response: Promise<unknown>, request: RpcRequest): void
> ```

That also solves for my use case, though less clear to me - I would expect `onResponse` to trigger once the promise resolves, not when the promise was created.

Could do the best of both worlds:

```ts
onRequest(request: RpcRequest, response: Promise<unknown>): void
onResponse(response: unknown, request: RpcRequest): void
```

```ts
const responsePromise = request(args);
onRequest?.(args, responsePromise);

const response = await responsePromise;
onResponse?.(response, args);

return response;
```

Ultimately, my use case is wanting to know when a tx is triggered so I can show details about the pending tx (even before it hits the mem pool) and ultimately update those details as the tx goes through its lifecycle (hits the mempool, gets mined, etc.)

**Comment by jxom** [comment]
Closing as this PR is becoming stale.

**Pull request closed without merging**